packages/material-ui-lab/src/TreeNode/TreeNode.js

```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import MenuItem from '@material-ui/core/MenuItem';
+import ListItemIcon from '@material-ui/core/ListItemIcon';
+import Typography from '@material-ui/core/Typography';
+import IconButton from '@material-ui/core/IconButton';
+import Collapse from '@material-ui/core/Collapse';
+import { makeStyles } from '@material-ui/styles';
+import useTreeState from '../TreeView/useTreeState';
+
+const useStyles = makeStyles(theme => {
+  const basePadding = theme.spacing(1);
+  const indentSize = 2;
+
+  return {
+    root: {
+      paddingLeft: props => basePadding + theme.spacing(props.depth * indentSize),
+    },
+    title: {
+      flex: 1,
+    },
+    stateIconContainer: {
+      flex: `0 0 ${30 + theme.spacing(1)}px`,
+      paddingRight: theme.spacing(1),
+    },
+    iconRoot: {
+      marginRight: theme.spacing(1),
+    },
+  };
+});
+
+function TreeNode(props) {
+  const { children, collapseIcon, depth, expandIcon, icon, id, title: titleProp, ...other } = props;
+  const classes = useStyles(props);
+  const { isExpanded, toggle, icons } = useTreeState();
```

Comment:
Can we put the depth into the context as well?

Reply:
We can, but I was trying to decide whether we would want a context per Node with children?

Reply:
I though every node should keep its own expanded state. Otherwise the whole tree would rerender if you expand some deep nested node. 

It just looks like a missed opportunity if we already have a component tree and then reimplement that. As far as I can tell a node should only care about its depth and if it is expanded. The rest is already implemented in react.

I would be as radical as completely removing the special root implementation. It should only exists as nice API for default values.

Reply:
@eps1lon How would you solve the controlling issue? For instance: https://ant.design/components/tree/#components-tree-demo-search

Reply:
I’m in the process of moving to nested ul > li > ul

Reply:
> @eps1lon How would you solve the controlling issue?

Probably with some `connect`-like wrapper.